src/20250114_AaveV3Ethereum_GSMsMigrationToGSM4626/AaveV3Ethereum_GSMsMigrationToGSM4626_20250114.t.sol

```diff
@@ -0,0 +1,410 @@
+    IGsm(proposal.NEW_GSM_USDT()).buyAsset(500e6, address(this));
+  }
+
+  function _checkRolesConfig(IGsm gsm) internal view {
+    // DAO permissions
+    assertTrue(
+      gsm.hasRole(bytes32(0), GovernanceV3Ethereum.EXECUTOR_LVL_1),
+      'Executor is not admin'
+    );
+    assertTrue(
+      gsm.hasRole(gsm.SWAP_FREEZER_ROLE(), GovernanceV3Ethereum.EXECUTOR_LVL_1),
+      'Executor is not swap freezer'
+    );
+    assertTrue(
+      gsm.hasRole(gsm.CONFIGURATOR_ROLE(), GovernanceV3Ethereum.EXECUTOR_LVL_1),
+      'Executor is not configurator'
+    );
+    // No need to be liquidator or token rescuer at the beginning
+    assertFalse(gsm.hasRole(gsm.LIQUIDATOR_ROLE(), GovernanceV3Ethereum.EXECUTOR_LVL_1));
+    assertFalse(gsm.hasRole(gsm.TOKEN_RESCUER_ROLE(), GovernanceV3Ethereum.EXECUTOR_LVL_1));
+
+    // Deployer does not have permissions
+    address deployer = 0x99C7A4A4Ab99882C422eF777b182eBda204D5B02;
+    assertFalse(gsm.hasRole(bytes32(0), deployer), 'Deployer cannot be admin');
+    assertFalse(gsm.hasRole(gsm.SWAP_FREEZER_ROLE(), deployer), 'Deployer cannot be swap freezer');
+    assertFalse(gsm.hasRole(gsm.CONFIGURATOR_ROLE(), deployer), 'Deployer cannot be configurator');
+    assertFalse(gsm.hasRole(gsm.LIQUIDATOR_ROLE(), deployer), 'Deployer cannot be liquidator');
+    assertFalse(
+      gsm.hasRole(gsm.TOKEN_RESCUER_ROLE(), deployer),
+      'Deployer cannot be token rescuer'
+    );
+  }
+
+  function _mockAssetPrice(address priceOracle, address asset, uint256 price) internal {
+    vm.mockCall(
+      priceOracle,
+      abi.encodeWithSelector(IPriceOracle.getAssetPrice.selector, asset),
+      abi.encode(price)
+    );
+  }
+
+  struct GsmConfig {
+    uint256 sellFee;
+    uint256 buyFee;
+    uint256 exposureCap;
+    bool isFrozen;
+    bool isSeized;
+    bool freezerCanUnfreeze;
+    uint256 freezeLowerBound;
+    uint256 freezeUpperBound;
+    uint256 unfreezeLowerBound;
+    uint256 unfreezeUpperBound;
+  }
+
+  function _checkGsmConfig(
+    IGsm gsm,
+    address underlying,
+    IOracleSwapFreezer freezer,
+    GsmConfig memory config
+  ) internal view {
+    assertEq(gsm.UNDERLYING_ASSET(), underlying, 'wrong underlying asset');
+    assertLt(gsm.getAvailableUnderlyingExposure(), config.exposureCap, 'wrong exposure cap');
+    assertEq(gsm.getIsFrozen(), config.isFrozen, 'wrong freeze state');
+    assertEq(gsm.getIsSeized(), config.isSeized, 'wrong seized state');
+
+    // Fee Strategy
+    IFeeStrategy feeStrategy = IFeeStrategy(gsm.getFeeStrategy());
+    assertEq(feeStrategy.getSellFee(10000), config.sellFee, 'wrong sell fee');
+    assertEq(feeStrategy.getBuyFee(10000), config.buyFee, 'wrong buy fee');
+
+    // Price Strategy
+    IFixedPriceStrategy4626 priceStrategy = IFixedPriceStrategy4626(gsm.PRICE_STRATEGY());
+    assertApproxEqAbs(priceStrategy.getAssetPriceInGho(1e6, true), 1 ether, 0.15 ether);
+    assertApproxEqAbs(priceStrategy.getGhoPriceInAsset(1 ether, true), 1e6, 0.15e6);
```

Comment:
do we need such big delta?

Reply:
@miguelmtzinf the price is not 1:1 now because of the value of static a tokens, could do strict checks, was just saying static a is ~1.15:1 to GHO currently.

Reply:
i d change this assert (or add a new one), using the 4626 conversion functions to calculate the accurate price of the asset in GHO terms (and the reverse)